File: articles/puberty-embodiment-and-adolescent-transition/c/puberty_summary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define MAX_LINE 16384
int main(int argc, char *argv[]) {
  if (argc < 2) { fprintf(stderr, "usage: %s data/puberty_embodiment_panel.csv\n", argv[0]); return 1; }
  FILE *fp = fopen(argv[1], "r"); if (!fp) { perror("open"); return 1; }
  char line[MAX_LINE]; fgets(line, sizeof(line), fp);
  int score=-1,protective=-1,stigma=-1,body=-1,peer=-1,col=0;
  char *tok=strtok(line,",");
  while(tok){ tok[strcspn(tok,"\r\n")]=0; if(!strcmp(tok,"adjustment_score")) score=col; if(!strcmp(tok,"protective_context")) protective=col; if(!strcmp(tok,"current_stigma")) stigma=col; if(!strcmp(tok,"current_body_concern")) body=col; if(!strcmp(tok,"current_peer_comparison")) peer=col; tok=strtok(NULL,","); col++; }
  long n=0; double ss=0,cc=0,xx=0,bb=0,pp=0;
  while(fgets(line,sizeof(line),fp)){ int c=0; char *f=strtok(line,","); double a=0,d=0,x=0,b=0,p=0; while(f){ if(c==score)a=atof(f); if(c==protective)d=atof(f); if(c==stigma)x=atof(f); if(c==body)b=atof(f); if(c==peer)p=atof(f); f=strtok(NULL,","); c++; } n++; ss+=a; cc+=d; xx+=x; bb+=b; pp+=p; }
  fclose(fp);
  printf("Rows analyzed: %ld\nMean adjustment_score: %.4f\nMean protective_context: %.4f\nMean current_stigma: %.4f\nMean current_body_concern: %.4f\nMean current_peer_comparison: %.4f\n", n, ss/n, cc/n, xx/n, bb/n, pp/n);
  return 0;
}

File: articles/puberty-embodiment-and-adolescent-transition/go/puberty_summary.go
package main
import ("encoding/csv"; "fmt"; "log"; "os"; "strconv")
func idx(h []string, name string) int { for i,v := range h { if v == name { return i } }; return -1 }
func main() {
  if len(os.Args) < 2 { log.Fatal("usage: go run puberty_summary.go data/puberty_embodiment_panel.csv") }
  f, err := os.Open(os.Args[1]); if err != nil { log.Fatal(err) }; defer f.Close()
  rows, err := csv.NewReader(f).ReadAll(); if err != nil { log.Fatal(err) }
  h := rows[0]; s:=idx(h,"adjustment_score"); c:=idx(h,"protective_context"); x:=idx(h,"current_stigma"); b:=idx(h,"current_body_concern"); p:=idx(h,"current_peer_comparison")
  n:=0; ss:=0.0; cc:=0.0; xx:=0.0; bb:=0.0; pp:=0.0
  for _, r := range rows[1:] { a,_:=strconv.ParseFloat(r[s],64); d,_:=strconv.ParseFloat(r[c],64); e,_:=strconv.ParseFloat(r[x],64); g,_:=strconv.ParseFloat(r[b],64); q,_:=strconv.ParseFloat(r[p],64); n++; ss+=a; cc+=d; xx+=e; bb+=g; pp+=q }
  fmt.Printf("Rows analyzed: %d\nMean adjustment_score: %.4f\nMean protective_context: %.4f\nMean current_stigma: %.4f\nMean current_body_concern: %.4f\nMean current_peer_comparison: %.4f\n", n, ss/float64(n), cc/float64(n), xx/float64(n), bb/float64(n), pp/float64(n))
}
